WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/ProductSelectorViewModelTests.swift

```diff
@@ -116,15 +116,27 @@ final class ProductSelectorViewModelTests: XCTestCase {
         XCTAssertEqual(viewModel.syncStatus, .empty)
     }
 
+    @MainActor
     func test_sync_status_updates_as_expected_when_products_are_synced() {
         // Given
-        let viewModel = ProductSelectorViewModel(siteID: sampleSiteID, storageManager: storageManager, stores: stores)
-        stores.whenReceivingAction(ofType: ProductAction.self) { action in
+        let mockStorageManager = MockStorageManager()
+        let mockStores = MockStoresManager(sessionManager: .testingInstance)
+        let viewModel = ProductSelectorViewModel(
+            siteID: sampleSiteID,
+            storageManager: mockStorageManager,
+            stores: mockStores
+        )
+        var syncStatusSpy: [ProductSelectorViewModel.SyncStatus] = []
+
+        mockStores.whenReceivingAction(ofType: ProductAction.self) { action in
             switch action {
             case let .synchronizeProducts(_, _, _, _, _, _, _, _, _, _, onCompletion):
-                XCTAssertEqual(viewModel.syncStatus, .firstPageSync)
-                let product = Product.fake().copy(siteID: self.sampleSiteID, purchasable: true)
-                self.insert(product)
+                if let syncStatus = viewModel.syncStatus {
+                    syncStatusSpy.append(syncStatus)
+                }
+                let readOnlyProduct = Product.fake().copy(siteID: self.sampleSiteID, purchasable: true)
+                let product = mockStorageManager.viewStorage.insertNewObject(ofType: StorageProduct.self)
+                product.update(with: readOnlyProduct)
```

Comment:
I didn't extract this bit into a different function because this is the only test that requires we're specific about which storage we're inserting objects to, due to its async nature it is relegated to a different thread that the one used for the test class, so if we use the `insert(_:)` helper method and run the whole test class it will crash as we're attempting to mutate a context queue from a different context queue.

Reply:
❓  I was wondering why the insertion of the product isn't needed before the PR changes. what are the changes in the view model that affect the storage behavior?

> Due to its async nature it is relegated to a different thread that the one used for the test class, so if we use the insert(_:) helper method and run the whole test class it will crash as we're attempting to mutate a context queue from a different context queue.

How about waiting for the completion of `ProductAction.synchronizeProducts` by `waitFor` in the When block, `viewModel.sync(pageNumber: 1, pageSize: 25, onCompletion: { _ in })` below? Then the test case will only finish after the sync is complete.

Reply:
> I was wondering why the insertion of the product isn't needed before the PR changes. what are the changes in the view model that affect the storage behavior?

I may be off here, but as far as I could see when debugging, this is due to the removal of `.dropFirst()` in the `.sink`. From the previous existing implementation, we seem to call the former `configureProductSearch()` twice on initialization, and with this calling resynchronize to storage. Using `.dropFirst()` had the side-effect of removing one of the sync calls, hiding the issue of attempting to write and read from the same thread (at least in unit tests)

Here's the slack thread for context: p1681094050104429-slack-CGPNUU63E

Reply:
> How about waiting for the completion of ProductAction.synchronizeProducts by waitFor in the When block, viewModel.sync(pageNumber: 1, pageSize: 25, onCompletion: { _ in }) below? Then the test case will only finish after the sync is complete.

Just to clarify: Are you referring to add the assertions inside the waitFor block? Such as:

```
        // When
        let _: Bool = waitFor { completion in
            viewModel.sync(pageNumber: 1, pageSize: 25, onCompletion: { sync in
                if let syncStatus = viewModel.syncStatus {
                    syncStatusSpy.append(syncStatus)
                }
                completion(sync)
            })
            XCTAssertEqual(syncStatusSpy, [.firstPageSync, .results])
            XCTAssertEqual(viewModel.syncStatus, .results)
        }
```

Rather than 

```
        // When
        let _: Bool = waitFor { completion in
            viewModel.sync(pageNumber: 1, pageSize: 25, onCompletion: { sync in
                if let syncStatus = viewModel.syncStatus {
                    syncStatusSpy.append(syncStatus)
                }
                completion(sync)
            })
        }
        XCTAssertEqual(syncStatusSpy, [.firstPageSync, .results])
        XCTAssertEqual(viewModel.syncStatus, .results)
```

Or you meant just regarding when we intercept the sync status?